### ISSUE TITLE
Class/forName considered as function name (proposal: say that it must be a class symbol)

### ISSUE THREAD

**Issue by arichiardi** [clj-kondo/clj-kondo#2277]
- [X] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

**version** 
2024.02.12

**platform**
linux - native binary

**problem**
I have the following in a file (for extending XTDB's handler):

```clojure
(extend-protocol pedestal.servlet-interceptor/WriteableBody
           (Class/forName "[B") ;; This errors out
           (write-body-to-stream [^InputStream input-stream ^OutputStream output-stream]
             (with-open [out output-stream] (.write out ^bytes input-stream)))
           juxt.clojars_mirrors.muuntaja.v0v6v8.muuntaja.protocols.StreamableResponse
           (write-body-to-stream [^InputStream input-stream ^OutputStream output-stream]
             (with-open [out output-stream] ((.f input-stream) ^OutputStream out))))
```

The highlighter line is reported as:

```shell
src/web2/server.clj:53:27: error: Function name must be simple symbol but got: "[B"
```
I don't think it's a configuration error and it only happens with the latest version of `clj-kondo`.

**expected behavior**

It should not error out as above.

**Comment by borkdude** [comment]
You can't officially use `(Class/forName "[B")` with `extend-protocol`.
Only class names are allowed, not expressions. That it works is accidental. In this case it's better to write the `extend` call or simply. I can give links where this is discussed later on. One of the issues with doing what you do above is that it becomes unreliable when you have multiple expressions like that.

**Comment by arichiardi** [comment]
I see, did not know that!

> better to write the extend call or simply (?)

I'd like to know what "simply" is here cause I like simple :)

**Comment by borkdude** [comment]
That was a typo. Writing `extend` is simple.

**Comment by borkdude** [comment]
In Clojure 1.12 you will be able to write `(extend-protocol IFoo byte* ...)`

**Comment by borkdude** [comment]
Here is the relevant JIRA: https://clojure.atlassian.net/browse/CLJ-1381
Slack discussion: https://clojurians.slack.com/archives/CHY97NXE2/p1677266201633939

**Comment by borkdude** [comment]
I think we could improve the error message to say that the class name must be a simple symbol

**Comment by borkdude** [comment]
I'll leave this issue open until CLJ-2836 is resolved.

**Comment by borkdude** [comment]
Clojure 1.12 now allows you to write:

```
(extend-protocol IFoo byte/1 ...)
```
